google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionClient.java

```diff
@@ -0,0 +1,287 @@
+      Preconditions.checkArgument(sessionCount > 0, "sessionCount must be > 0");
+      this.channelHint = channelHint;
+      this.sessionCount = sessionCount;
+      this.consumer = consumer;
+    }
+
+    @Override
+    public void run() {
+      List<SessionImpl> sessions = null;
+      int remainingSessionsToCreate = sessionCount;
+      while (remainingSessionsToCreate > 0) {
+        try {
+          sessions = internalBatchCreateSessions(remainingSessionsToCreate, channelHint);
+        } catch (Throwable t) {
+          consumer.onSessionCreateFailure(t, remainingSessionsToCreate);
+          break;
+        }
+        int numActuallyCreated = sessions.size();
+        if (numActuallyCreated == 0) {
+          // No sessions returned by the server. Give up creation to avoid an infinite loop.
+          consumer.onSessionCreateFailure(
+              newSpannerException(ErrorCode.UNKNOWN, "Server did not return any sessions"),
+              remainingSessionsToCreate);
+          break;
+        }
+        for (SessionImpl session : sessions) {
+          consumer.onSessionReady(session);
+        }
+        remainingSessionsToCreate -= numActuallyCreated;
+      }
+    }
+  }
+
+  /**
+   * Callback interface to be used for BatchCreateSessions. When sessions become available or
+   * session creation fails, one of the callback methods will be called.
+   */
+  static interface SessionConsumer {
+    /** Called when a session has been created and is ready for use. */
+    void onSessionReady(SessionImpl session);
+
+    /**
+     * Called when an error occurred during session creation. The createFailureForSessionCount
+     * indicates the number of sessions that could not be created, so that the consumer knows how
+     * many sessions it should still expect.
+     */
+    void onSessionCreateFailure(Throwable t, int createFailureForSessionCount);
+  }
+
+  private final SpannerImpl spanner;
+  private final ExecutorFactory<ScheduledExecutorService> executorFactory;
+  private final ScheduledExecutorService executor;
+  private final DatabaseId db;
+
+  @GuardedBy("this")
+  private volatile long sessionChannelCounter;
+
+  SessionClient(
+      SpannerImpl spanner,
+      DatabaseId db,
+      ExecutorFactory<ScheduledExecutorService> executorFactory) {
+    this.spanner = spanner;
+    this.db = db;
+    this.executorFactory = executorFactory;
+    this.executor = executorFactory.get();
+  }
+
+  @Override
+  public void close() {
+    executorFactory.release(executor);
+  }
+
+  /** Create a single session. */
+  SessionImpl createSession() {
+    // The sessionChannelCounter could overflow, but that will just flip it to Integer.MIN_VALUE,
+    // which is also a valid channel hint.
+    final Map<SpannerRpc.Option, ?> options;
+    synchronized (this) {
+      options = optionMap(SessionOption.channelHint(sessionChannelCounter++));
+    }
+    Span span = SpannerImpl.tracer.spanBuilder(SpannerImpl.CREATE_SESSION).startSpan();
```

Comment:
should we also add tracing for everythig around BatchCreateSession..
If the traces do have any logging, please add numRequested & numReceived session counts which will greatly help with debugging.

Reply:
I've added tracing for BatchCreateSessions on two levels:
1. The `BatchCreateSessionsRunnable` creates a scoped span for the entire session creation, including an annotation indicating the number of sessions being requested.
1. The `BatchCreateSessions` rpc call has its own child span that also includes the number of requested and actually returned sessions.

Reply:
SGTM !